tsl/src/chunk.c

```diff
@@ -186,3 +206,889 @@ chunk_invoke_drop_chunks(Oid relid, Datum older_than, Datum older_than_type, boo
+		 * example, if doing a concurrent DROP TABLE on one of the chunks will
+		 * lead to deadlock because it grabs locks on all dependencies before
+		 * dropping.
+		 *
+		 * However, for now we won't do that because that requires scanning
+		 * pg_depends and concurrent operations will probably fail anyway if
+		 * we remove the objects. We might as well fail with a deadlock.
+		 */
+		LOCKMODE lockmode =
+			(lock_upgrade == MERGE_LOCK_ACCESS_EXCLUSIVE) ? AccessExclusiveLock : ExclusiveLock;
+
+		rel = try_table_open(relid, lockmode);
+
+		/* Check if the table actually exists. If not, it could have been
+		 * deleted in a concurrent merge. */
+		if (rel == NULL)
+			ereport(ERROR,
+					(errcode(ERRCODE_UNDEFINED_TABLE),
+					 errmsg("relation does not exist"),
+					 errdetail("The relation with OID %u might have been removed "
+							   "by a concurrent merge or other operation.",
+							   relid)));
+
+		if (rel->rd_rel->relkind == '\0')
+			ereport(ERROR,
+					(errcode(ERRCODE_UNDEFINED_TABLE),
+					 errmsg("relation with OID %u does not exist", relid)));
+
+		if (rel->rd_rel->relkind != RELKIND_RELATION)
+			ereport(ERROR,
+					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+					 errmsg("cannot merge non-table relations")));
+
+		/* Only owner is allowed to merge */
+		if (!object_ownercheck(RelationRelationId, relid, GetUserId()))
+			aclcheck_error(ACLCHECK_NOT_OWNER,
+						   get_relkind_objtype(rel->rd_rel->relkind),
+						   get_rel_name(relid));
+
+		/* Remember the relpersistence until processing next rel */
+		relpersistence = relpers;
+
+		if (relpersistence != '\0' && relpersistence != relpers)
+			ereport(ERROR,
+					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+					 errmsg("cannot merge relations with different persistence")));
+
+		/* Lock toast table to prevent it from being concurrently vacuumed */
+		if (rel->rd_rel->reltoastrelid)
+			LockRelationOid(rel->rd_rel->reltoastrelid, lockmode);
+
+		/*
+		 * Check for active uses of the relation in the current transaction,
+		 * including open scans and pending AFTER trigger events.
+		 */
+		CheckTableNotInUse(rel, "merge_chunks");
+
+		if (IsSystemRelation(rel))
+			ereport(ERROR,
+					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					 errmsg("cannot merge system catalog relations")));
+
+		/*
+		 * Find the chunk corresponding to the relation for final checks. Done
+		 * after locking the chunk relation because scanning for the chunk
+		 * will grab locks on other objects, which might otherwise lead to
+		 * deadlocks during concurrent merges instead of more helpful messages
+		 * (like chunk does not exist because it was merged).
+		 */
+		chunk = ts_chunk_get_by_relid(relid, false);
+
+		if (NULL == chunk)
+			ereport(ERROR,
+					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+					 errmsg("can only merge hypertable chunks")));
+
+		if (chunk->fd.osm_chunk)
+			ereport(ERROR,
+					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED), errmsg("cannot merge OSM chunks")));
```

Comment:
We should add some tests for it. See chunk_utils_internal.sql

Reply:
Added test case.